File: framework/languages/node/samples/Bingo.Ts/Server/Play/Domain/Bingo/bingo-room-game.ts
const { BingoGame } = require('./bingo-game');
const { BingoRoomStatus } = require('./bingo-room-models');

type BingoActor = {
  actorId: string;
  displayName: string;
};

type BingoRoomSettings = {
  requiredPlayers: number;
  drawDeck: number[];
};

type BingoPlayerSeat = {
  actor: BingoActor;
  seat: number;
  card: any | null;
  isHost: boolean;
};

class BingoRoomGame {
  [key: string]: any;
  constructor(roomId: string, settings: BingoRoomSettings) {
    this.roomId = roomId;
    this.settings = {
      requiredPlayers: settings.requiredPlayers,
      drawDeck: [...settings.drawDeck]
    };
    this.status = BingoRoomStatus.waitingForPlayers;
    this.players = [];
    this.game = new BingoGame(this.settings.drawDeck);
  }

  join(actor: BingoActor): { joined: boolean; player: BingoPlayerSeat; started: boolean } {
    const existing = this.players.find((player) => player.actor.actorId === actor.actorId);
    if (existing !== undefined) {
      return { joined: false, player: existing, started: false };
    }
    if (this.status !== BingoRoomStatus.waitingForPlayers || this.players.length >= this.settings.requiredPlayers) {
      throw new Error(`Room ${this.roomId} cannot accept more players.`);
    }
    const player = {
      actor,
      seat: this.players.length,
      card: null,
      isHost: this.players.length === 0
    };
    this.players.push(player);
    let started = false;
    if (this.players.length === this.settings.requiredPlayers) {
      this.status = BingoRoomStatus.running;
      started = true;
    }
    return { joined: true, player, started };
  }

  submitCard(actorId: string, cardNumbers: number[]): void {
    if (this.status !== BingoRoomStatus.running) {
      throw new Error(`Room ${this.roomId} is not running.`);
    }
    const player = this.requirePlayer(actorId);
    player.card = this.game.submitCard({ actorId, card: player.card }, cardNumbers);
  }

  canDraw(): boolean {
    return this.status === BingoRoomStatus.running
      && this.game.canDraw(this.players.map((player) => ({ actorId: player.actor.actorId, card: player.card })), this.settings.requiredPlayers);
  }

  drawNext(): { number: number; drawSeq: number; finished: boolean } | null {
    if (!this.canDraw()) {
      return null;
    }
    const drawn = this.game.drawNext(this.players.map((player) => ({ actorId: player.actor.actorId, card: player.card })));
    if (drawn !== null && drawn.finished) {
      this.status = BingoRoomStatus.finished;
    }
    return drawn;
  }

  snapshot(): any {
    return {
      roomId: this.roomId,
      status: this.status,
      hostActorId: this.players[0]?.actor.actorId ?? null,
      canStart: false,
      drawSeq: this.game.drawnNumbers.length,
      lastDrawnNumber: this.game.lastDrawnNumber(),
      drawnNumbers: [...this.game.drawnNumbers],
      players: this.players.map((player) => ({
        actorId: player.actor.actorId,
        displayName: player.actor.displayName,
        seat: player.seat,
        isHost: player.isHost,
        card: player.card === null ? [] : [...player.card.numbers],
        marks: player.card === null ? [] : [...player.card.marks],
        completedLines: player.card === null ? 0 : player.card.completedLines()
      })),
      winners: [...this.game.winners]
    };
  }

  requirePlayer(actorId: string): BingoPlayerSeat {
    const player = this.players.find((entry) => entry.actor.actorId === actorId);
    if (player === undefined) {
      throw new Error(`Actor '${actorId}' has not joined room '${this.roomId}'.`);
    }
    return player;
  }
}

export { BingoRoomGame };

File: framework/languages/dotnet/samples/Bingo/Server/Play/Domain/Bingo/BingoRoomGame.cs
using Bingo.Shared.Contracts;

namespace Bingo.Server.Play.Domain.Bingo;

internal sealed class BingoRoomGame(string roomId, BingoRoomSettings settings)
{
    private readonly List<BingoRoomPlayer> _players = [];
    private BingoRoomSettings _settings = settings;
    private BingoGame? _game;

    public string Status { get; private set; } = BingoRoomStatus.WaitingForPlayers;

    public bool IsReadyToDraw => _game?.IsReadyToDraw == true
                                 && Status == BingoRoomStatus.Running;

    public void ApplySettings(BingoRoomSettings newSettings)
    {
        if (newSettings.RequiredPlayers <= 0)
        {
            throw new InvalidOperationException("Bingo room requires at least one player.");
        }

        if (newSettings.MaxDrawNumber <= 0)
        {
            throw new InvalidOperationException("Bingo room requires at least one draw number.");
        }

        _settings = newSettings;
        _game = null;
    }

    public BingoGameChange JoinPlayer(string actorId, string displayName)
    {
        var existing = _players.FirstOrDefault(player => player.ActorId == actorId);
        if (existing is not null)
        {
            return new BingoGameChange(Snapshot(), []);
        }

        if (Status != BingoRoomStatus.WaitingForPlayers || _players.Count >= _settings.RequiredPlayers)
        {
            throw new InvalidOperationException($"Room {roomId} cannot accept more players.");
        }

        var player = new BingoRoomPlayer(actorId, displayName, _players.Count);
        _players.Add(player);

        var events = new List<BingoGameEvent>();
        var joinedState = Snapshot();
        events.AddRange(PlayerJoinedEvents(player, joinedState));

        if (_players.Count == _settings.RequiredPlayers)
        {
            Status = BingoRoomStatus.Running;
            _game = new BingoGame(_settings, _players.Select(static roomPlayer => roomPlayer.ActorId).ToArray());
            events.AddRange(EventsForExistingPlayers(
                BingoRoomEventKind.GameStarted,
                Snapshot(),
                player.ActorId));
        }

        return new BingoGameChange(Snapshot(), events);
    }

    public BingoGameChange SubmitCard(string actorId, BingoCard card)
    {
        if (Status != BingoRoomStatus.Running)
        {
            throw new InvalidOperationException($"Room is not accepting cards. status={Status}");
        }

        RequireGame().SubmitCard(actorId, card);
        return new BingoGameChange(Snapshot(), []);
    }

    public BingoGameChange DrawNextNumber()
    {
        if (Status != BingoRoomStatus.Running)
        {
            return new BingoGameChange(Snapshot(), [], ShouldStopDrawTimer: true);
        }

        var result = RequireGame().DrawNextNumber();
        if (result.IsFinished)
        {
            Status = BingoRoomStatus.Finished;
        }

        var state = Snapshot();
        var events = new List<BingoGameEvent>();
        if (result.Number is { } number)
        {
            events.AddRange(NumberDrawnEvents(state, number));
        }

        if (Status == BingoRoomStatus.Finished)
        {
            events.AddRange(EventsForAll(BingoRoomEventKind.GameEnded, state));
        }

        return new BingoGameChange(
            state,
            events,
            ShouldStopDrawTimer: Status == BingoRoomStatus.Finished);
    }

    public BingoRoomState Snapshot()
    {
        var hostActorId = _players.Count == 0 ? string.Empty : _players[0].ActorId;
        var game = _game?.Snapshot() ?? new BingoGameSnapshot(0, null, [], []);
        var state = new BingoRoomState
        {
            RoomId = roomId,
            Status = Status,
            HostActorId = hostActorId,
            CanStart = Status == BingoRoomStatus.WaitingForPlayers && _players.Count == _settings.RequiredPlayers,
            DrawSeq = game.DrawSeq,
            DrawnNumbers = { game.DrawnNumbers },
            Players = { _players.Select(player => player.ToState(hostActorId, _game)) },
            Winners = { game.Winners },
        };
        if (game.LastDrawnNumber is { } lastDrawnNumber)
        {
            state.LastDrawnNumber = lastDrawnNumber;
        }

        return state;
    }

    private IReadOnlyList<BingoGameEvent> PlayerJoinedEvents(
        BingoRoomPlayer joined,
        BingoRoomState state)
    {
        return _players
            .Where(player => !string.Equals(player.ActorId, joined.ActorId, StringComparison.Ordinal))
            .Select(player => new BingoGameEvent(
                BingoRoomEventKind.PlayerJoined,
                player.ActorId,
                state,
                joined.ActorId,
                joined.DisplayName,
                joined.Seat,
                joined.Seat == 0))
            .ToArray();
    }

    private IReadOnlyList<BingoGameEvent> NumberDrawnEvents(
        BingoRoomState state,
        int number)
    {
        return _players
            .Select(player => new BingoGameEvent(
                BingoRoomEventKind.NumberDrawn,
                player.ActorId,
                state,
                DrawnNumber: number))
            .ToArray();
    }

    private IReadOnlyList<BingoGameEvent> EventsForAll(
        BingoRoomEventKind kind,
        BingoRoomState state)
    {
        return _players
            .Select(player => new BingoGameEvent(kind, player.ActorId, state))
            .ToArray();
    }

    private IReadOnlyList<BingoGameEvent> EventsForExistingPlayers(
        BingoRoomEventKind kind,
        BingoRoomState state,
        string joinedActorId)
    {
        return _players
            .Where(player => !string.Equals(player.ActorId, joinedActorId, StringComparison.Ordinal))
            .Select(player => new BingoGameEvent(kind, player.ActorId, state))
            .ToArray();
    }

    private BingoGame RequireGame()
    {
        return _game ?? throw new InvalidOperationException("Bingo game has not started.");
    }

    private sealed record BingoRoomPlayer(
        string ActorId,
        string DisplayName,
        int Seat)
    {
        public BingoPlayerState ToState(string hostActorId, BingoGame? game)
        {
            var card = game?.CardSnapshot(ActorId) ?? BingoCardSnapshot.Empty;
            return new BingoPlayerState
            {
                ActorId = ActorId,
                DisplayName = DisplayName,
                Seat = Seat,
                IsHost = string.Equals(ActorId, hostActorId, StringComparison.Ordinal),
                Card = { card.Numbers },
                Marks = { card.Marks },
                CompletedLines = card.CompletedLines,
            };
        }
    }
}
